### ISSUE TITLE
Pulsar binder integ test failure on JDK 21

### ISSUE THREAD

**Issue by sobychacko** [spring-cloud/spring-cloud-stream#2915]
[PulsarBinderIntegrationTests#binderAndBindingPropsAreAppliedAndRespected](https://github.com/spring-cloud/spring-cloud-stream/blob/main/binders/pulsar-binder/spring-cloud-stream-binder-pulsar/src/test/java/org/springframework/cloud/stream/binder/pulsar/PulsarBinderIntegrationTests.java#L79) is failing on JDK 21 because of some issues with the property [max-pending-messages](https://github.com/spring-cloud/spring-cloud-stream/blob/main/binders/pulsar-binder/spring-cloud-stream-binder-pulsar/src/test/java/org/springframework/cloud/stream/binder/pulsar/PulsarBinderIntegrationTests.java#L91). Please investigate, why this property has issues on JDK 21, but not on JDK 17.

**Comment by onobc** [comment]
## TL;DR
The error has been there from inception and is only noticed because the `ProducerConfigurationData.setMaxPendingMessages` and `ProducerConfigurationData.setMaxPendingMessagesAcrossPartitions` are called in different order in Java17 and 21. 

If the failing test would have instead set only the `maxPendingMessagesAcrossPartitions` we would have seen the same issue in Java17. 

## Details
The binder allows users to configure the underlying producer used by the binder. 

The config props are layered from `base -> binder -> binding` and we only include those properties that have changed from the defaults. 

When hydrating the ProducerConfigurationData with the modified props, the `setMaxPendingMessagesAcrossPartitions` validates that it is `>= the maxPendingMessages`. 

In the case where only the `maxPendingMessages` is set, the `maxPendingMessagesAcrossPartitions` has a default of 0. 
If the `setMaxPendingMessagesAcrossPartitions` is called after `setMaxPendingMessages` then the validation will fail.

In Java17, the serialized JSON and merged set of props were put into a HashMap which just so happened to result in `maxPendingMessagesAcrossPartitions` being called before `maxPendingMessages` and we did not notice the issue in the test.

In Java21, the resulting props in the HashMap just so happen to result in `maxPendingMessages` being called before `maxPendingMessagesAcrossPartitions` and we see the validation error.

## Solution
I think we should only be limiting the producer props set by the user for the base producer props (the same ones set by Spring Boot). For the extended producer props (those supported only by the binder and not boot) we should also include the default values whether or not the user overrides the default. The reason for this is that the extended props defaults have not be set prior to the binder using the producer. The base props have (via Spring Boot auto-config).
